### PR TITLE
Fixed Form Sample Crash

### PR DESCRIPTION
**Pull request by alexruperez**
There was a crash filling and viewing results in the form sample.

### CODE REVIEW

**Comment by jwe-apple** [comment]
Could you please give clear instructions to reproduce the crash? Maybe you have, but I've not seen them.

File: samples/ORKCatalog/ORKCatalog/TaskListRow.swift [hunk 1]
```diff
@@ -642,7 +642,7 @@ enum TaskListRow: Int, Printable {
         
         // A first field, for entering an integer.
         let formItem01Text = NSLocalizedString("Field01", comment: "")
-        let formItem01 = ORKFormItem(identifier: Identifier.FormItem01.rawValue, text: formItem01Text, answerFormat: ORKAnswerFormat.integerAnswerFormatWithUnit(nil))
+        let formItem01 = ORKFormItem(identifier: Identifier.FormItem01.rawValue, text: formItem01Text, answerFormat: ORKAnswerFormat.integerAnswerFormatWithUnit(""))
```
Comment (jwe-apple) [inline samples/ORKCatalog/ORKCatalog/TaskListRow.swift]: This isn't the right fix. Instead, the unit in ORKNumericQuestionResult should be marked nullable.

File: ResearchKit/Common/ORKTableContainerView.m [hunk 2]
```diff
@@ -280,7 +280,7 @@ - (CGSize)keyboardIntersectionSizeFromNotification:(NSNotification *)notificatio
 
 - (void)scrollCellVisible:(UITableViewCell *)cell animated:(BOOL)animated {
     
-    if (cell == nil) {
+    if (cell == nil || [cell respondsToSelector:@selector(bounds)] == NO) {
```
Comment (jwe-apple) [inline ResearchKit/Common/ORKTableContainerView.m]: If something is calling this method with an object that is not either nil or a UITableViewCell, then that is the bug. Do you have a stack trace for the crash you hit?

**Comment by mattio** [comment]
I may have run into the same issue so here is what was dumped to the console for me. 

Repro Steps:

1: Open the Form sample in the ORKCatalog app.
2: Type any text in Field01.
3: Tap to type in Field02. That will trigger the crash consistently for me on OS X 10.10.3, Xcode 6.3 (6D570).

```
2015-04-15 16:24:45.092 ORKCatalog[2363:60740] -[ORKTaskViewController showViewController:goForward:animated:] 1010 <ORKTaskViewController: 0x7ff1d9f31c90> <ORKFormStepViewController: 0x7ff1d9d59d70>
2015-04-15 16:24:45.092 ORKCatalog[2363:60740] -[ORKStepViewController viewWillAppear:] 149 <ORKFormStepViewController: 0x7ff1d9d59d70>
2015-04-15 16:24:45.093 ORKCatalog[2363:60740] __63-[ORKTaskViewController showViewController:goForward:animated:]_block_invoke668 1027 <ORKTaskViewController: 0x7ff1d9f31c90> <ORKFormStepViewController: 0x7ff1d9d59d70>
2015-04-15 16:24:45.104 ORKCatalog[2363:60740] -[ORKFormItemPickerCell setDefaultAnswer:] 879 (null)
2015-04-15 16:24:45.105 ORKCatalog[2363:60740] -[ORKFormItemPickerCell setDefaultAnswer:] 879 (null)
2015-04-15 16:24:45.118 ORKCatalog[2363:60740] __66-[ORKTaskViewController requestHealthAuthorizationWithCompletion:]_block_invoke_2 567 Requesting health access
2015-04-15 16:24:45.118 ORKCatalog[2363:60740] -[ORKFormItemPickerCell setDefaultAnswer:] 879 (null)
2015-04-15 16:25:25.266 ORKCatalog[2363:60740] -[ORKTableContainerView view:hasFirstResponderOrTableViewCellContainingPoint:] 198 v=<ORKFormItemNumericCell: 0x7ff1d9ea1a10; baseClass = UITableViewCell; frame = (0 134; 320 60.5); autoresize = W; layer = <CALayer: 0x7ff1d9eb5db0>>
2015-04-15 16:25:25.318 ORKCatalog[2363:60740] -[ORKTableContainerView scrollCellVisible:animated:] 289 height=504.000000 insetsbottom=0.000000
2015-04-15 16:25:25.327 ORKCatalog[2363:60740] Can't find keyplane that supports type 4 for keyboard iPhone-Portrait-NumberPad; using 1730230351_Portrait_iPhone-Simple-Pad_Default
2015-04-15 16:25:28.841 ORKCatalog[2363:60740] -[ORKTableContainerView view:hasFirstResponderOrTableViewCellContainingPoint:] 198 v=<ORKFormItemPickerCell: 0x7ff1d9ecd590; baseClass = UITableViewCell; frame = (0 194.5; 320 60.5); autoresize = W; layer = <CALayer: 0x7ff1d9ecdf40>>
2015-04-15 16:25:28.981 ORKCatalog[2363:60740] -[ORKTableContainerView scrollCellVisible:animated:] 289 height=504.000000 insetsbottom=0.000000
2015-04-15 16:25:29.001 ORKCatalog[2363:60740] -[ORKTableContainerView scrollCellVisible:animated:] 289 height=504.000000 insetsbottom=216.000000
2015-04-15 16:25:29.002 ORKCatalog[2363:60740] -[ORKTableContainerView scrollCellVisible:animated:] 289 height=504.000000 insetsbottom=216.000000
2015-04-15 16:25:29.014 ORKCatalog[2363:60740] -[UIAlertController superview]: unrecognized selector sent to instance 0x7ff1d9de8c60
2015-04-15 16:25:29.017 ORKCatalog[2363:60740] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[UIAlertController superview]: unrecognized selector sent to instance 0x7ff1d9de8c60'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010c4b5c65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x000000010e020bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x000000010c4bd0ad -[NSObject(NSObject) doesNotRecognizeSelector:] + 205
	3   CoreFoundation                      0x000000010c41313c ___forwarding___ + 988
	4   CoreFoundation                      0x000000010c412cd8 _CF_forwarding_prep_0 + 120
	5   ResearchKit                         0x000000010c147ffe __62-[ORKTableContainerView animateLayoutForKeyboardNotification:]_block_invoke + 1566
	6   UIKit                               0x000000010cdcd422 +[UIView(UIViewAnimationWithBlocks) _setupAnimationWithDuration:delay:view:options:factory:animations:start:animationStateGenerator:completion:] + 473
	7   UIKit                               0x000000010cdcd677 +[UIView(UIViewAnimationWithBlocks) animateWithDuration:delay:options:animations:completion:] + 57
	8   ResearchKit                         0x000000010c14796a -[ORKTableContainerView animateLayoutForKeyboardNotification:] + 410
	9   ResearchKit                         0x000000010c14826a -[ORKTableContainerView keyboardFrameWillChange:] + 202
	10  CoreFoundation                      0x000000010c48554c __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
	11  CoreFoundation                      0x000000010c383a04 _CFXNotificationPost + 2484
	12  Foundation                          0x000000010c894968 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
	13  UIKit                               0x000000010d41be9b -[UIInputWindowController postStartNotifications:withInfo:] + 124
	14  UIKit                               0x000000010d41d832 __77-[UIInputWindowController moveFromPlacement:toPlacement:starting:completion:]_block_invoke576 + 375
	15  UIKit                               0x000000010cdcd422 +[UIView(UIViewAnimationWithBlocks) _setupAnimationWithDuration:delay:view:options:factory:animations:start:animationStateGenerator:completion:] + 473
	16  UIKit                               0x000000010cdcd638 +[UIView(UIViewAnimationWithBlocks) _animateWithDuration:delay:options:animations:start:completion:] + 53
	17  UIKit                               0x000000010cf02916 -[_UIViewControllerKeyboardAnimationStyle launchAnimation:afterStarted:completion:forHost:fromCurrentPosition:] + 553
	18  UIKit                               0x000000010d41d53d -[UIInputWindowController moveFromPlacement:toPlacement:starting:completion:] + 972
	19  UIKit                               0x000000010d4214c2 -[UIInputWindowController setInputViewSet:] + 599
	20  UIKit                               0x000000010d41cf6a -[UIInputWindowController performOperations:withAnimationStyle:] + 50
	21  UIKit                               0x000000010d1ffa0e -[UIPeripheralHost(UIKitInternal) setInputViews:animationStyle:] + 1054
	22  UIKit                               0x000000010ced866d -[UIResponder becomeFirstResponder] + 468
	23  UIKit                               0x000000010cfbe01d -[UIAlertController viewWillAppear:] + 98
	24  UIKit                               0x000000010ce88fa1 -[UIViewController _setViewAppearState:isAnimating:] + 487
	25  UIKit                               0x000000010ce62548 __56-[UIPresentationController runTransitionForCurrentState]_block_invoke + 581
	26  UIKit                               0x000000010cd7862c _applyBlockToCFArrayCopiedToStack + 314
	27  UIKit                               0x000000010cd784a6 _afterCACommitHandler + 533
	28  CoreFoundation                      0x000000010c3e8ca7 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	29  CoreFoundation                      0x000000010c3e8c00 __CFRunLoopDoObservers + 368
	30  CoreFoundation                      0x000000010c3dea33 __CFRunLoopRun + 1123
	31  CoreFoundation                      0x000000010c3de366 CFRunLoopRunSpecific + 470
	32  GraphicsServices                    0x0000000110109a3e GSEventRunModal + 161
	33  UIKit                               0x000000010cd54900 UIApplicationMain + 1282
	34  ORKCatalog                          0x000000010beedb57 main + 135
	35  libdyld.dylib                       0x000000010f955145 start + 1
	36  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by jwe-apple** [comment]
Ah, perfect. The bug is on line 384 of ORKTableViewContainer.m, where it casts the first responder as UIView. In this case it's an incorrect assumption. That block of code should only continue if firstResponderView is in fact a UIView subclass.

Happy to accept a revised version of this patch that addresses the root cause. Thanks for finding!

**Comment by alexruperez** [comment]
Revised version of this patch that addresses the root cause here: https://github.com/alexruperez/ResearchKit/commit/422d3afb4185278d3904b2109ec5b168e6285e45

Thanks @jwe-apple @mattio @davedelong

File: ResearchKit/Common/UIApplication+ResearchKit.m [hunk 3]
```diff
@@ -34,15 +34,20 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
...
     NSMutableArray *sender = [NSMutableArray new] ;
     [self sendAction:@selector(ork_updateCurrentFirstResponder:)
                                                to:nil
                                              from:sender
                                          forEvent:nil];
     
+    for (UIResponder *responder in sender) {
+        if ([responder isKindOfClass:[UIView class]]) {
+            return (UIView *)responder;
+        }
```
Comment (YuanZhu-apple) [inline ResearchKit/Common/UIApplication+ResearchKit.m]: We don't need to walk this array - the array only has one object. (see implementation of ork_updateCurrentFirstResponder:)

Why not just do:
id responder = [sender firstObject];
if (! [responder isKindOfClass:[UIView class]]) {
    responder = nil;
}
return responder;


This also would neatly avoid an early return.

**Comment by alexruperez** [comment]
@YuanZhu-apple nice solution, thanks!

Fixed at https://github.com/alexruperez/ResearchKit/commit/5de3d1ec8bbfe831d3ce5ccb9fe81d8beea436ba

**Pull request merged**